### PR TITLE
new lint: comparing unit types (fixes #201)

### LINKED ISSUE #201: Warn on comparisons involving the unit type
Because the unit type will only ever contain one value, any comparisons are trivial and are likely accidental. Superfluous semicolons at the end of scopes could easily cause this.

The following code compiles and prints "That's funny...":
```
fn main() {
    if {true;} == {false;} {
        println!("That's funny...");
    }
}
```

The `let_unit_value` lint would normally catch this, but it doesn't cover all cases and a separate lint seems easy enough to implement.

### PR DESCRIPTION
**Pull request by birkenfeld**

### CODE REVIEW

**Comment by llogiq** [comment]
What about `() > ()`? I would extend this to cover all comparison ops.

`syntax::ast_util::is_comparison_binop(BinOp_)` may be useful here. The expected result is `true` for `==`, `<=`, `>=`, false otherwise.

**Comment by birkenfeld** [comment]
Huh, I wouldn't have expected unit to even implement `Ord`... But I guess it's implemented for all tuples where all innner types are, which is definitely true for `()` :smile:

File: src/types.rs [hunk 1]
```diff
@@ -107,3 +108,32 @@ impl LintPass for LetPass {
...
+        lint_array!(UNIT_CMP)
+    }
+
+    fn check_expr(&mut self, cx: &Context, expr: &Expr) {
+        if let ExprBinary(ref cmp, ref left, _) = expr.node {
+            let op = cmp.node;
+            let sty = &cx.tcx.expr_ty(left).sty;
+            if *sty == ty::TyTuple(vec![]) && is_comparison_binop(op) {
+                let result = match op {
+                    BiEq | BiLe | BiNe => "true",
```
Comment (llogiq) [inline src/types.rs]: That should be `BiGe`, not `BiNe`.

File: src/types.rs [hunk 2]
```diff
@@ -107,3 +108,32 @@ impl LintPass for LetPass {
             |info| check_let_unit(cx, decl, info));
     }
 }
+
+declare_lint!(pub UNIT_CMP, Warn,
+              "using `==` or `!=` on unit values (which is always `true` or `false`, \
```
Comment (llogiq) [inline src/types.rs]: Should read "comparing unit values"

**Comment by llogiq** [comment]
Otherwise LGTM

File: src/types.rs (hunk 1 above)
Comment (birkenfeld) [inline src/types.rs]: :shame:

Comment (llogiq) [inline src/types.rs]: Happens to the best of us.

**Pull request merged**